### ISSUE TITLE
Force market exit after timeout when custom_stoploss exit order remains pending

### ISSUE THREAD

**Issue by sithu-go** [freqtrade/freqtrade#12175]
**Description**

Currently, when a stoploss is triggered and an exit order is placed with limit price, the bot continues to wait for some minutes for the order to be filled in some cases, even if market conditions have changed significantly. That causes me an expected loss. I don't want to use `stoploss_on_exchange`. 

I want to implement a custom stoploss logic where:

1. If stoploss is triggered, a stoploss order is placed (limit order).
2. If after 1 minute the exit order is still not executed (still pending), I want to force exit at market price immediately.

Please help me check my custom_stop_loss is ok.

```py
# pragma pylint: disable=missing-docstring, invalid-name, pointless-string-statement
# flake8: noqa: F401
# isort: skip_file
# --- Do not remove these imports ---
import numpy as np
import pandas as pd
from datetime import datetime, timedelta, timezone
from pandas import DataFrame
from typing import Dict, Optional, Union, Tuple

from freqtrade.strategy import (
    IStrategy,
    Trade,
    Order,
    PairLocks,
    informative,  # @informative decorator
    # Hyperopt Parameters
    BooleanParameter,
    CategoricalParameter,
    DecimalParameter,
    IntParameter,
    RealParameter,
    # timeframe helpers
    timeframe_to_minutes,
    timeframe_to_next_date,
    timeframe_to_prev_date,
    # Strategy helper functions
    merge_informative_pair,
    stoploss_from_absolute,
    stoploss_from_open,
    AnnotationType,
)

# --------------------------------
# Add your lib to import here
import talib.abstract as ta
from technical import qtpylib

slh: dict[str, dict] = {}

class BBandsRSI(IStrategy):
    # Strategy interface version
    INTERFACE_VERSION = 3

    # Optimal timeframe for the strategy.
    timeframe = "5m"

    # Can this strategy go short?
    can_short: bool = False

    use_custom_stoploss = True

    # Minimal ROI designed for the strategy.
    minimal_roi = {
        "60": 0.01,
        "30": 0.02,
        "0": 0.04
    }

    # Optimal stoploss designed for the strategy.
    stoploss = -0.10

    # Trailing stoploss
    trailing_stop = False

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = True

    # These values can be overridden in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False

    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 30

    # Strategy parameters
    buy_rsi = IntParameter(10, 40, default=30, space="buy")
    sell_rsi = IntParameter(60, 90, default=70, space="sell")

    # Optional order type mapping.
    order_types = {
        "entry": "limit",
        "exit": "limit",
        "stoploss": "market",
        "stoploss_on_exchange": False
    }

    # Optional order time in force.
    order_time_in_force = {
        "entry": "GTC",
        "exit": "GTC"
    }

    @property
    def plot_config(self):
        return {
            "main_plot": {
                # Add OHLC data to main plot
                "bb_upperband": {"color": "grey"},
                "bb_middleband": {"color": "red"},
                "bb_lowerband": {"color": "grey"},
            },
            "subplots": {
                # Subplots - each dict defines one additional plot
                "RSI": {
                    "rsi": {"color": "blue"},
                    "overbought": {"color": "red", "type": "line"},
                    "oversold": {"color": "green", "type": "line"}
                }
            }
        }

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # RSI
        dataframe["rsi"] = ta.RSI(dataframe, timeperiod=14)
        dataframe["overbought"] = 70
        dataframe["oversold"] = 30 

        # Bollinger Bands
        bollinger = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe["bb_lowerband"] = bollinger["lower"]
        dataframe["bb_middleband"] = bollinger["mid"]
        dataframe["bb_upperband"] = bollinger["upper"]
        dataframe["bb_percent"] = (
            (dataframe["close"] - dataframe["bb_lowerband"]) /
            (dataframe["bb_upperband"] - dataframe["bb_lowerband"])
        )
        dataframe["bb_width"] = (
            (dataframe["bb_upperband"] - dataframe["bb_lowerband"]) / dataframe["bb_middleband"]
        )

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe["rsi"] < self.buy_rsi.value) &  # RSI oversold
                (dataframe["close"] <= dataframe["bb_lowerband"]) &  # Price at lower BB
                (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "enter_long"] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe["rsi"] > self.sell_rsi.value) |  # RSI overbought
                (dataframe["close"] >= dataframe["bb_upperband"]) &  # Price at upper BB
                (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "exit_long"] = 1
 
        return dataframe
    
    # even after 1 minute after stoploss hit, if exit order is still pending, use market price to exit immediately
    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                            current_rate: float, current_profit: float, after_fill: bool, 
                            **kwargs) -> float | None:
        tradeID = trade.id
        
        if current_profit < self.stoploss:
            if tradeID in slh:
                # Stop loss already hit, check if > 1 minute
                when_hit = slh[tradeID]["when_hit"]
                if datetime.now() - when_hit > timedelta(minutes=1):

                    # sell at market price immediately
                    del slh[tradeID]
                    return {
                        "exit_reason": "force_stoploss",
                        "exit_type": "force_exit"   # ensures market order if configured
                    }
                return True
            else:
                slh[tradeID] = {
                    "is_stop_loss_hit": True,
                    "when_hit": datetime.now()
                }

        return None
```

**Comment by stash86** [comment]
Combine https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-order-timeout-rules with https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-order-price-rules.

You can't return Bool nor dict from custom_stoploss. If you get that from AI, please stop using AI.

**Comment by sithu-go** [comment]
what about `unfilledtimeout` in config
If i change to `exit` to 1, does that complete my requirement?
```json
    "unfilledtimeout": {
        "entry": 10,
        "exit": 1,
        "exit_timeout_count": 0,
        "unit": "minutes"
    },
```

**Comment by stash86** [comment]
You can try. Either way, you can't use both limit and market orders at the same time. Pick one. To have limit order behave as if it's market, give it a low(for sell) or high(for buy) enough prices

**Comment by sithu-go** [comment]
What will happen if I set exit to 1?
After one minute, does trade automatically exit?

**Comment by stash86** [comment]
after 1 min, the limit order canceled

**Comment by sithu-go** [comment]
Yes, But another limit order comes up based on roi and stoploss, right?

**Comment by xmatthias** [comment]
depends - if the signal is still active - yes.
if the price moved above the stoploss level (or below the ROI level) - no.
